File: wfm-develop mobile 07.05.2026/wfm/mobile/ios/WFMApp/Core/Models/ShiftTimeCalculator.swift
import Foundation

/// Утилита для расчета времени смен
///
/// Содержит все функции для форматирования и расчета времени:
/// - Определение опоздания на смену
/// - Парсинг времени начала/конца смены
/// - Форматирование сообщений о времени
struct ShiftTimeCalculator {

    // MARK: - Public Methods

    /// Определить, опаздывает ли пользователь на смену
    static func isShiftLate(_ shift: CurrentShift, currentTime: Date = Date()) -> Bool {
        guard let start = shiftStartDate(shift) else { return false }
        return currentTime > start
    }

    /// Получить дату и время начала смены
    static func shiftStartDate(_ shift: CurrentShift?) -> Date? {
        guard let shift = shift,
              let dateStr = shift.shiftDate,
              let timeStr = shift.startTime else { return nil }
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        let timeOnly = String(timeStr.prefix(8))
        return formatter.date(from: "\(dateStr) \(timeOnly)")
    }

    /// Получить дату и время конца смены
    static func shiftEndDate(_ shift: CurrentShift?) -> Date? {
        guard let shift = shift, let timeStr = shift.endTime else { return nil }
        let timeOnly = String(timeStr.prefix(8))
        let dateStr: String
        if let d = shift.shiftDate {
            dateStr = d
        } else if let openedAt = shift.openedAt {
            let dateFmt = DateFormatter()
            dateFmt.dateFormat = "yyyy-MM-dd"
            dateStr = dateFmt.string(from: openedAt)
        } else {
            return nil
        }
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return formatter.date(from: "\(dateStr) \(timeOnly)")
    }

    /// Форматировать количество минут до начала смены
    ///
    /// Примеры:
    /// - 0 мин → "Начинается сейчас"
    /// - 30 мин → "Начнется через 30 мин"
    /// - 90 мин → "Начнется через 1 ч 30 мин"
    static func formatMinutesUntil(_ minutes: Int) -> String {
        if minutes == 0 { return "Начинается сейчас" }
        let h = minutes / 60
        let m = minutes % 60
        if h == 0 { return "Начнется через \(m) мин" }
        if m == 0 { return "Начнется через \(h) ч" }
        return "Начнется через \(h) ч \(m) мин"
    }

    /// Форматировать количество минут опоздания
    ///
    /// Примеры:
    /// - 0 мин → "Вы опаздываете"
    /// - 15 мин → "Вы опаздываете на 15 мин"
    /// - 75 мин → "Вы опаздываете на 1 ч 15 мин"
    static func formatMinutesLate(_ minutes: Int) -> String {
        if minutes == 0 { return "Вы опаздываете" }
        let h = minutes / 60
        let m = minutes % 60
        if h == 0 { return "Вы опаздываете на \(m) мин" }
        if m == 0 { return "Вы опаздываете на \(h) ч" }
        return "Вы опаздываете на \(h) ч \(m) мин"
    }

    /// Форматировать оставшееся время до конца смены
    ///
    /// Примеры:
    /// - 0 мин → "Смена заканчивается"
    /// - 45 мин → "До конца смены 45 мин"
    /// - 120 мин → "До конца смены 2 ч"
    static func formatMinutesLeft(_ minutes: Int) -> String {
        if minutes == 0 { return "Смена заканчивается" }
        let h = minutes / 60
        let m = minutes % 60
        if h == 0 { return "До конца смены \(m) мин" }
        if m == 0 { return "До конца смены \(h) ч" }
        return "До конца смены \(h) ч \(m) мин"
    }
}

File: wfm-develop mobile 07.05.2026/wfm/mobile/android/app/src/main/java/com/beyondviolet/wfm/core/models/ShiftTimeCalculator.kt
package com.beyondviolet.wfm.core.models

import kotlinx.datetime.toJavaInstant
import java.text.SimpleDateFormat
import java.util.Locale

/**
 * Утилита для расчета времени смен
 *
 * Содержит все функции для форматирования и расчета времени:
 * - Определение опоздания на смену
 * - Парсинг времени начала/конца смены
 * - Форматирование сообщений о времени
 */
object ShiftTimeCalculator {

    /**
     * Определить, опаздывает ли пользователь на смену
     */
    fun isShiftLate(shift: CurrentShift, now: Long = System.currentTimeMillis()): Boolean {
        return (shiftStartMillis(shift) ?: return false) < now
    }

    /**
     * Получить время начала смены в миллисекундах
     */
    fun shiftStartMillis(shift: CurrentShift): Long? {
        val dateStr = shift.shiftDate ?: return null
        val timeStr = shift.startTime ?: return null
        return try {
            SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
                .parse("$dateStr ${timeStr.take(8)}")?.time
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Получить время конца смены в миллисекундах
     */
    fun shiftEndMillis(shift: CurrentShift): Long? {
        val timeStr = shift.endTime ?: return null
        val dateStr = shift.shiftDate
            ?: shift.openedAt?.toJavaInstant()
                ?.atZone(java.time.ZoneId.systemDefault())
                ?.toLocalDate()
                ?.toString()
            ?: return null
        return try {
            SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
                .parse("$dateStr ${timeStr.take(8)}")?.time
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Форматировать количество минут до начала смены
     *
     * Примеры:
     * - 0 мин → "Начинается сейчас"
     * - 30 мин → "Начнется через 30 мин"
     * - 90 мин → "Начнется через 1 ч 30 мин"
     */
    fun formatMinutesUntil(minutes: Int): String {
        if (minutes == 0) return "Начинается сейчас"
        val h = minutes / 60
        val m = minutes % 60
        return when {
            h == 0 -> "Начнется через $m мин"
            m == 0 -> "Начнется через $h ч"
            else -> "Начнется через $h ч $m мин"
        }
    }

    /**
     * Форматировать количество минут опоздания
     *
     * Примеры:
     * - 0 мин → "Вы опаздываете"
     * - 15 мин → "Вы опаздываете на 15 мин"
     * - 75 мин → "Вы опаздываете на 1 ч 15 мин"
     */
    fun formatMinutesLate(minutes: Int): String {
        if (minutes == 0) return "Вы опаздываете"
        val h = minutes / 60
        val m = minutes % 60
        return when {
            h == 0 -> "Вы опаздываете на $m мин"
            m == 0 -> "Вы опаздываете на $h ч"
            else -> "Вы опаздываете на $h ч $m мин"
        }
    }

    /**
     * Форматировать оставшееся время до конца смены
     *
     * Примеры:
     * - 0 мин → "Смена заканчивается"
     * - 45 мин → "До конца смены 45 мин"
     * - 120 мин → "До конца смены 2 ч"
     */
    fun formatMinutesLeft(minutes: Int): String {
        if (minutes == 0) return "Смена заканчивается"
        val h = minutes / 60
        val m = minutes % 60
        return when {
            h == 0 -> "До конца смены $m мин"
            m == 0 -> "До конца смены $h ч"
            else -> "До конца смены $h ч $m мин"
        }
    }
}
